pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -844,13 +908,101 @@ func (n *NetworkPolicyController) deleteService(old interface{}) {
 	}
 	defer n.heartbeat("deleteService")
 
-	klog.V(2).Infof("Processing Service %s/%s DELETE event", service.Namespace, service.Name)
+	klog.V(2).InfoS("Processing Service DELETE event", "serviceNamespace", service.Namespace, "serviceName", service.Name)
 	// Find all internal Group keys which refers to this Service.
 	groupKeySet := n.filterInternalGroupsForService(service)
 	// Enqueue internal groups to its queue for group processing.
 	for group := range groupKeySet {
 		n.enqueueInternalGroup(group)
 	}
+
+	// If this Service doesn't have a selector, there won't be an Endpoints DELETE event
+	// triggered by this Service DELETE event. So find all ACNPs that are impacted by
+	// this Service and reprocess them here. Otherwise, an Endpoints DELETE event will
+	// reprocess those ACNPs.
+	if service.Spec.Selector == nil {
+		affectedACNPsByService := n.filterACNPsByService(service.Namespace, service.Name)
+		for _, acnp := range affectedACNPsByService {
+			n.reprocessACNP(acnp)
+		}
+	}
+}
+
+// addEndpoints receives Endpoints ADD events and reprocesses the
+// ClusterNetworkPolicy with `toService` referred to the Service which uses this
+// Endpoints as backend.
+func (n *NetworkPolicyController) addEndpoints(obj interface{}) {
+	defer n.heartbeat("addEndpoints")
+	endpoints := obj.(*v1.Endpoints)
+	klog.V(2).InfoS("Processing Endpoints ADD event", "endpointsNamespace", endpoints.Namespace, "endpointsName", endpoints.Name)
+	// Find all ACNPs that are impacted by this Endpoints and reprocess them.
+	// Because Service and its Endpoints use the same Namespace and Name, pass
+	// Namespace and Name of Endpoints into filterACNPsByService
+	affectedACNPsByService := n.filterACNPsByService(endpoints.Namespace, endpoints.Name)
+	for _, acnp := range affectedACNPsByService {
+		n.reprocessACNP(acnp)
```

Comment:
m a bt concerned that endpoint add reprocesses all acnps and ths s repeated for all endpoints .. there seems to be a lot of duplicate processing and that too within a handler and not workers

Reply:
If Service has a selector then the Service won't trigger reprocesses.